torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+      // vector after erasing.
+      auto clear_left_of = [&find_outer_it,
+                            &outer_pos,
+                            &find_inner_it,
+                            &inner_pos,
+                            &ids,
+                            &merge](decltype(find_outer_it) it) {
+        if (it != ids.begin()) {
+          ids.erase(ids.begin(), it);
+        }
+        find_outer_it = std::find(ids.begin(), ids.end(), merge->outer());
+        outer_pos = find_outer_it == ids.end()
+            ? ids.size()
+            : std::distance(ids.begin(), find_outer_it);
+
+        find_inner_it = std::find(ids.begin(), ids.end(), merge->inner());
+        inner_pos = find_inner_it == ids.end()
+            ? ids.size()
+            : std::distance(ids.begin(), find_inner_it);
+      };
+
+      // Dry run to fill state
+      clear_left_of(ids.begin());
+
+      // Check if the input domains of the merge are contiguous in the mapped
+      // domain.
+      if (find_outer_it == ids.end() && find_inner_it != ids.end()) {
+        // Outer dimension was not found, but inner dimension was. Must assume
+        // everything to the left of inner is not contiguously merged.
+        //
+        // Clear left of inner
+        clear_left_of(find_inner_it);
+      } else if (find_outer_it != ids.end() && find_inner_it == ids.end()) {
+        // Inner dimension was not found, outer and anything left of outer are
+        // definitely not contiguous.
+        //
+        // Clear outer and left of outer
+        clear_left_of(find_outer_it + 1);
+        continue;
+      } else if (find_outer_it == ids.end() && find_inner_it == ids.end()) {
+        // Nothing mapped, just continue
         continue;
       }
-      auto maybe_dimension_size =
-          expression_evaluator.evaluate(root_id->extent());
+
+      if (find_outer_it != ids.end() && find_inner_it != ids.end()) {
+        // Both outer and inner mapped.
+        if (outer_pos >= inner_pos) {
+          // Make sure outer is outside inner, otherwise neither could be part
+          // of a continuous mapping. There are cases where we could have
+          // reversible operations e.g.:
+          //    [id{3} id{5} id{6}] -> merge(1, 0)
+          // -> [id{5*3} id{6}] -> split(0, 5)
+          // -> [id{5} id{3} id{6}] -> transpose(0, 1)
+          // -> [id{3} id{5} id{6}]
+          // However we don't try and capture cases like this correcly, we'd
+          // just reduce this down to only the iter domain of size 6 mapping.
+          //
+          // Clear outer and left of outer
+          clear_left_of(find_outer_it + 1);
+          continue;
+        }
+
+        // Find the position inner would have to have to be considered ordered
+        // relative to outer
+        auto pos_after_outer = outer_pos + 1;
+        for (; pos_after_outer < ids.size(); pos_after_outer++) {
+          if (ids[pos_after_outer]->isBroadcast()) {
```

Comment:
Thanks, I think it might be necessary, as it's fine to have an inner position as broadcast, if that's not the inner position than we can just ignore the broadcast, but if it is the inner position we can pick it up.